### PR TITLE
XW-955 | Add userID param to GA and u param for Internal tracking

### PR DESCRIPTION
**Pull request by vforge**
https://wikia-inc.atlassian.net/browse/XW-955

### CODE REVIEW

File: front/common/modules/Trackers/UniversalAnalytics.js [hunk 1]
```diff
@@ -103,11 +105,13 @@ class UniversalAnalytics {
 	 * @returns {void}
 	 */
 	initAccount(trackerName, domain) {
-		const options = {
-			name: '',
-			allowLinker: true,
-			sampleRate: this.accounts[trackerName].sampleRate
-		};
+		const gaUserIdHash = M.prop('gaUserIdHash', ''),
```
Comment (rogatty) [inline front/common/modules/Trackers/UniversalAnalytics.js]: The second parameter makes it a setter, not getter.

File: front/common/modules/Trackers/UniversalAnalytics.js [hunk 2]
```diff
@@ -103,11 +105,13 @@ class UniversalAnalytics {
...
+			options = {
+				name: '',
+				allowLinker: true,
+				sampleRate: this.accounts[trackerName].sampleRate,
+				userId: (gaUserIdHash.length > 0 ? gaUserIdHash : undefined)
```
Comment (rogatty) [inline front/common/modules/Trackers/UniversalAnalytics.js]: What's wrong with `null`?

File: server/config/localSettings.base.js [hunk 3]
```diff
@@ -171,6 +171,8 @@ const localSettings = {
 	mediawikiDomain: process.env.MEDIAWIKI_DOMAIN || null,
 	// Special salt for accepting HTML from MediaWiki for /editor_preview/
 	mwPreviewSalt: process.env.MW_PREVIEW_SALT,
+	// Special salt for calculating GA userID
+	gaUserSalt: process.env.GA_USER_SALT,
```
Comment (rogatty) [inline server/config/localSettings.base.js]: We probably should update the JSDoc too.

File: front/common/tests/unit/modules/Trackers/UniversalAnalytics.js [hunk 4]
```diff
@@ -128,8 +129,14 @@ QUnit.module('mercury/modules/Trackers/UniversalAnalytics', function (hooks) {
 		assert.strictEqual(queueCount('require'), 1);
 		assert.strictEqual(queueCount('ads.require'), 1);
 
-		assert.ok(queueContains(['create', '123', 'auto', {name: '', allowLinker: true, sampleRate: 10}]));
-		assert.ok(queueContains(['create', '789', 'auto', {name: 'ads', allowLinker: true, sampleRate: 100}]));
+		assert.ok(queueContains([
```
Comment (hakubo) [inline front/common/tests/unit/modules/Trackers/UniversalAnalytics.js]: can you add tests for scenario with userId !== null?

File: server/app/facets/operations/prepareCuratedContentData.js [hunk 5]
```diff
@@ -71,7 +71,13 @@ export default function prepareCuratedContentData(request, curatedContentPageDat
 		result.localSettings.weppy.samplingRate = parseInt(request.query.buckySampling, 10) / 100;
 	}
 
-	result.userId = request.state.wikicitiesUserID ? request.state.wikicitiesUserID : 0;
+	if (request.auth.isAuthenticated) {
```
Comment (hakubo) [inline server/app/facets/operations/prepareCuratedContentData.js]: this looks like copy and paste, any chance to abstract it away?

File: server/app/facets/showApplication.js [hunk 6]
```diff
@@ -58,7 +59,13 @@ export default function showApplication(request, reply, wikiVariables) {
 	context.server = Utils.createServerData(localSettings, wikiDomain);
 	context.queryParams = Utils.parseQueryParams(request.query, []);
 	context.localSettings = localSettings;
-	context.userId = request.auth.isAuthenticated ? request.auth.credentials.userId : 0;
+	if (request.auth.isAuthenticated) {
```
Comment (hakubo) [inline server/app/facets/showApplication.js]: oh so its used 3 times :(

Comment (rogatty) [inline server/app/facets/showApplication.js]: Yeah, that's why we have https://wikia-inc.atlassian.net/browse/XW-608

File: server/app/lib/Utils.js [hunk 7]
```diff
@@ -359,3 +360,13 @@ export function getHtmlTitle(wikiVariables, displayTitle) {
...
 	return htmlTitleTemplate.replace('$1 - ', '');
 }
+
+/**
+ * Get md5 of string
+ *
+ * @param {string} data
+ * @returns {string}
+ */
+export function md5(data) {
```
Comment (hakubo) [inline server/app/lib/Utils.js]: data does not sound like a string

also 
it's relatively free to have another file, can we not make Utils a module that contains everything?
It's already almost 400 lines of code that should be split into files

File: server/app/facets/showApplication.js (hunk 6 above)
Comment (hakubo) [inline server/app/facets/showApplication.js]: ticket is there, that's good.
but can we not add to the spaghetti?

File: front/common/tests/unit/modules/Trackers/UniversalAnalytics.js (hunk 4 above)
Comment (vforge) [inline front/common/tests/unit/modules/Trackers/UniversalAnalytics.js]: see https://github.com/Wikia/mercury/pull/1890/files#diff-1deb03267718e6c5c9b057c4fbae39b6R142 below

**Comment by rogatty** [comment]
LGTM
